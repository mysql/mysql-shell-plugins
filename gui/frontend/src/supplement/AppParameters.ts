/*
 * Copyright (c) 2025, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import type { IAppParameters } from "./RequisitionTypes.js";

/** A set of values passed to the application via URL parameters and a couple other application wide values. */
export const appParameters: Map<string, string> & IAppParameters = new Map<string, string>();

/**
 * Determines if the app is embedded and wires up message handlers for sending and receiving wrapper messages.
 * It also converts URL parameters to a map, for consumption in the app.
 */
export const parseAppParameters = (): void => {
    if (typeof window !== "undefined") {
        const queryParts = window.location.search.substring(1).split("&");
        queryParts.forEach((part) => {
            const elements = part.split("=");
            if (elements.length > 1) {
                appParameters.set(elements[0], elements[1]);

                if (elements[0] === "app") {
                    appParameters.embedded = true;
                } else if (elements[0] === "fontSize") {
                    const fontSize = parseInt(elements[1], 10);
                    if (!isNaN(fontSize)) {
                        appParameters.fontSize = fontSize;
                    }
                } else if (elements[0] === "editorFontSize") {
                    const fontSize = parseInt(elements[1], 10);
                    if (!isNaN(fontSize)) {
                        appParameters.editorFontSize = fontSize;
                    }
                }
            }
        });
    }

    if (process.env.NODE_ENV === "test") {
        appParameters.testsRunning = true;
    } else if (process.env.NODE_ENV === "development") {
        appParameters.inDevelopment = true;
    }

    // Test Explorer tests also run under VS Code control, but must not be treated as embedded.
    if (process.env.VSCODE_PID !== undefined && process.env.JEST_WORKER_ID === undefined) {
        appParameters.inExtension = true;
    }
};

