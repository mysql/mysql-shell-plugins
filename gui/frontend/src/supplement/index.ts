/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { mysqlInfo, sqliteInfo } from "../app-logic/RdbmsInfo";
import { DBDataType, IColumnInfo, IDictionary } from "../app-logic/Types";
import { DBType } from "./ShellInterface";

// Commonly used data types and functions.

/**
 * These are the supported languages in the code editor.
 * This includes our mixed language (msg), which combines SQL, Python, JS and TS in one editor.
 */
export type EditorLanguage = (
    "text" | "typescript" | "javascript" | "mysql" | "sql" | "python" | "json" | "markdown" | "msg" | "xml" | "ini"
);

export interface ITextRange {
    readonly startLine: number;
    readonly startColumn: number;
    readonly endLine: number;
    readonly endColumn: number;
}

export interface IExecutionContext {
    id: string;
    code: string;
    codeLength: number;
    language: EditorLanguage;
    isSQLLike: boolean;
    linkId?: number;
}

export interface ISqlPageRequest {
    context: IExecutionContext;
    oldRequestId: string;
    page: number;
    sql: string;
}

export interface IRunQueryRequest {
    data: IDictionary;
    query: string;
    parameters: Array<[string, string]>;
    linkId: number;
}

export interface IScriptRequest {
    /** A unique ID to identify the script in this request. */
    scriptId: string;

    /** The language of the script. */
    language: EditorLanguage;

    name?: string;
    content: string;
}

/** Used to request the creation of a new script in the extension. */
export interface INewScriptRequest {
    /** The id of the page (connection tab) to open the new script in. */
    page: string;

    /** The language to use for the new script file. */
    language: EditorLanguage;

    /** Optional content to use for the new script file. */
    content?: string;
}

export { Stack } from "./Stack";

export type WorkerExecutorType<T> = (
    onResult?: ((taskId: number, value: T) => void),
    onError?: ((taskId: number, reason?: unknown) => void)
) => void;

// A callback with a promise-like callback function.
export interface IThenableCallback<T> {
    then: (onResult?: ((taskId: number, value: T) => void)) => IThenableCallback<T>;
    catch: (onError?: ((taskId: number, reason?: unknown) => void)) => void;
}

// This conditional type enforces type widening on generic parameter types.
export type ValueType<T> = T extends string
    ? string
    : T extends number
        ? number
        : T extends boolean
            ? boolean
            : T extends undefined
                ? undefined
                : [T] extends [unknown]
                    ? T
                    : object;

/**
 * Generates a generic column info record for each raw column returned from a request. The format of the
 * raw columns depends on the used RDBMS.
 *
 * @param dbType Determines for which DB type the conversion happens (MySQL, SQLite and so on).
 * @param rawColumns Column info as returned by the backend.
 * @param useName Use column name not index.
 *
 * @returns A list of columns with RDBMS-agnostic details.
 */
export const generateColumnInfo = (dbType: DBType, rawColumns?: Array<{ name: string; type: string; length: number }>,
    useName?: boolean): IColumnInfo[] => {
    if (!rawColumns) {
        return [];
    }

    const dataTypes = dbType === DBType.MySQL ? mysqlInfo.dataTypes : sqliteInfo.dataTypes;

    return rawColumns.map((entry, index) => {
        let type;
        if (entry.type.toLowerCase() === "bytes") {
            // For now, use length to switch between binary and blob
            if (entry.length < 256) {
                type = DBDataType.Binary;
            } else {
                type = DBDataType.Blob;
            }
        } else {
            const foundType = dataTypes.get(entry.type.toLowerCase());
            type = foundType ? foundType.type : DBDataType.Unknown;
        }

        return {
            title: entry.name,
            field: useName ? entry.name : String(index),
            dataType: {
                type,
            },
        };
    });
};

/**
 * Converts the given rows into a dictionary format as required by the UI, if it isn't already in that format.
 *
 * @param columns Column informations to find column IDs.
 * @param rows The rows to convert.
 *
 * @returns The converted rows.
 */
export const convertRows = (columns: IColumnInfo[], rows?: unknown[]): IDictionary[] => {
    if (!rows || rows.length === 0) {
        return [];
    }

    if (Array.isArray(rows[0])) {
        // Data is array of arrays of fields.
        const convertedRows: IDictionary[] = [];
        rows.forEach((value): void => {
            const row: IDictionary = {};

            const entry = value as unknown[];
            columns.forEach((column: IColumnInfo, columnIndex: number): void => {
                row[column.field] = entry[columnIndex];
            });
            convertedRows.push(row);
        });

        return convertedRows;
    }

    return rows as IDictionary[];
};

