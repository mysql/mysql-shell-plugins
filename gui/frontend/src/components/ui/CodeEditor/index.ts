/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { languages, editor } from "monaco-editor/esm/vs/editor/editor.api";
import { IStatementSpan } from "../../../parsing/parser-common";

import { IExecuteResultReference } from "../../../script-execution";
import { EditorLanguage } from "../../../supplement";

export type {
    IMarkdownString, IRange, IPosition, IDisposable, CancellationToken, Thenable,
} from "monaco-editor/esm/vs/editor/editor.api";

export {
    editor as Monaco, languages, KeyMod, KeyCode, Selection, Range, Position, Uri,
} from "monaco-editor/esm/vs/editor/editor.api";

export {
    ContextKeyExpr,
} from "monaco-editor/esm/vs/platform/contextkey/common/contextkey";

export type ProviderResult<T> = languages.ProviderResult<T>;

export type CompletionList = languages.CompletionList;
export type CompletionItem = languages.CompletionItem;
export type CompletionItemLabel = languages.CompletionItemLabel;
export type CompletionContext = languages.CompletionContext;

export type Hover = languages.Hover;

export type SignatureHelpContext = languages.SignatureHelpContext;
export type SignatureHelpResult = languages.SignatureHelpResult;
export type SignatureHelp = languages.SignatureHelp;
export type SignatureInformation = languages.SignatureInformation;
export type ParameterInformation = languages.ParameterInformation;

export type DocumentHighlight = languages.DocumentHighlight;

export type DocumentSymbol = languages.DocumentSymbol;

export type Definition = languages.Definition;
export type LocationLink = languages.LocationLink;

export type ReferenceContext = languages.ReferenceContext;
export type Location = languages.Location;

export type SelectionRange = languages.SelectionRange;

export type FormattingOptions = languages.FormattingOptions;
export type TextEdit = languages.TextEdit;
export type WorkspaceTextEdit = languages.WorkspaceTextEdit;

export type FoldingRange = languages.FoldingRange;
export type FoldingContext = languages.FoldingContext;

export type WorkspaceEdit = languages.WorkspaceEdit;
export type RenameLocation = languages.RenameLocation;
export type Rejection = languages.Rejection;

export type TypescriptWorker = languages.typescript.TypeScriptWorker;

export type ICodeEditorViewState = editor.ICodeEditorViewState;

export interface ICodeEditorOptions {
    tabSize?: number;
    indentSize?: number;
    insertSpaces?: boolean;

    defaultEOL?: "LF" | "CRLF";
    trimAutoWhitespace?: boolean;
}

export interface IExecutionContextState {
    /** The start line for this context. */
    start: number;

    /** The end line for this context. */
    end: number;

    /** The language used in the context. */
    language: EditorLanguage;

    /** Optionally: an attached result (reference). */
    result?: IExecuteResultReference;

    /** The height of the result pane. Only considered if the pane is not maximized. */
    currentHeight?: number;

    /** The index of the active set, in a multi set result. */
    currentSet?: number;

    /** When true no editor is shown and the result pane takes the entire space. */
    maximizeResultPane?: boolean;

    /** Stored statement ranges. */
    statements: IStatementSpan[];
}

// A statement consists of its text and its relative position within its block/context.
export interface IStatement {
    text: string;

    offset: number;  // The character offset of the statement in the containing model.
    line: number;    // The line number of the statement in the containing model.
    column: number;  // Ditto for the column.
}
