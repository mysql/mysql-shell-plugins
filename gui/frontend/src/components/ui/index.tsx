/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import React from "react";

import { ICodeEditorOptions } from "./CodeEditor";

export const defaultEditorOptions: ICodeEditorOptions = {
    tabSize: 4,
    indentSize: 4,
    insertSpaces: true,

    defaultEOL: "LF",
    trimAutoWhitespace: true,
};

export * from "./Component/Component";
export * from "./Container/Container";

export * from "./ActivityBar/ActivityBar";
export * from "./ActivityBar/ActivityBarItem";

export * from "./Button/Button";
export * from "./Checkbox/Checkbox";
export * from "./Radiobutton/Radiobutton";
export * from "./Dialog/Dialog";
export * from "./Icon/Icon";
export * from "./Codicon";
export * from "./Image/Image";
export * from "./Popup/Popup";
export * from "./Portal/Portal";

export * from "./Grid/Grid";
export * from "./Grid/GridCell";

export * from "./Label/Label";
export * from "./Input/Input";
export * from "./Search/Search";
export * from "./Message/Message";

export * from "./SplitContainer/SplitContainer";

export * from "./Accordion/Accordion";
export * from "./Accordion/AccordionItem";

export * from "./Tabview/Tabview";
export * from "./Toggle/Toggle";
export * from "./Divider/Divider";

export * from "./List/List";
export * from "./List/DynamicList";

export * from "./Selector/Selector";
export * from "./Selector/SelectorItem";

export * from "./Menu/Menu";
export * from "./Menu/MenuBar";
export * from "./Menu/MenuItem";

export * from "./Dropdown/Dropdown";

export * from "./Breadcrumb/Breadcrumb";
export * from "./Statusbar/Statusbar";

export * from "./TreeGrid/TreeGrid";

export * from "./ColorPicker/ColorPopup";
export * from "./ColorPicker/ColorField";

export * from "./Slider/Slider";
export * from "./Toolbar/Toolbar";
export * from "./TagInput/TagInput";
export * from "./UpDown/UpDown";
export * from "./Group/Group";
export * from "./Calendar/Calendar";

export * from "./FrontPage/FrontPage";
export * from "./MessageToast/MessageToast";
export * from "./FileSelector/FileSelector";
export * from "./AboutBox/AboutBox";
export * from "./ProgressIndicator/ProgressIndicator";

export * from "./ConnectionTile/ConnectionTile";
export * from "./SessionTile/SessionTile";
export * from "./BrowserTile/BrowserTile";

export * from "./StatusMark/StatusMark";

/**
 * Creates a synthetic event from a native DOM event, for interaction of DOM events with React components.
 *
 * @param event The DOM event to use.
 *
 * @returns A new synthetic event instance.
 */
export const createSyntheticEvent = <T extends Element, E extends Event>(event: E): React.SyntheticEvent<T, E> => {
    let isDefaultPrevented = false;
    let isPropagationStopped = false;
    const preventDefault = (): void => {
        isDefaultPrevented = true;
        event.preventDefault();
    };
    const stopPropagation = (): void => {
        isPropagationStopped = true;
        event.stopPropagation();
    };

    return {
        nativeEvent: event,
        currentTarget: event.currentTarget as EventTarget & T,
        target: event.target as EventTarget & T,
        bubbles: event.bubbles,
        cancelable: event.cancelable,
        defaultPrevented: event.defaultPrevented,
        eventPhase: event.eventPhase,
        isTrusted: event.isTrusted,
        preventDefault,
        stopPropagation,
        timeStamp: event.timeStamp,
        type: event.type,

        isDefaultPrevented: (): boolean => { return isDefaultPrevented; },
        isPropagationStopped: (): boolean => { return isPropagationStopped; },
        persist: (): void => { /**/ },
    };
};
