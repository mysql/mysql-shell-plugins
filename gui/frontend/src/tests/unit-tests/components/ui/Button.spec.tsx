/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import image from "../../../../assets/images/close.svg";

import React from "react";
import { mount, shallow } from "enzyme";
import { act } from "preact/test-utils";

import { Button, Image, Container, Orientation, ComponentSize, IButtonProperties } from "../../../../components/ui";
import { requisitions } from "../../../../supplement/Requisitions";
import { eventMock } from "../../__mocks__/MockEvents";
import { snapshotFromWrapper } from "../../test-helpers";

let clicked = false;

const buttonClick = (): void => {
    clicked = true;
};

const requestButtonClick = (): Promise<boolean> => {
    clicked = true;

    return Promise.resolve(true);
};

const sleepAsync = (callback: (flag: boolean) => void): ReturnType<typeof setImmediate> => {
    return setImmediate(() => { callback(clicked); });
};

describe("Button component tests", (): void => {

    beforeEach(() => {
        clicked = false;
    });

    afterAll(() => {
        requisitions.unregister(undefined, requestButtonClick);
    });

    it("Test button click", async () => {
        const component = shallow(
            <Button onClick={buttonClick}>
                Test button
            </Button>,
        );
        expect(component).toBeTruthy();
        expect(component.text()).toEqual("Test button");

        const click = (component.props() as IButtonProperties).onClick;
        await act(() => {
            click?.(eventMock, { id: "1" });
        });

        expect(clicked).toEqual(true);
    });

    it("Test button commands", (done) => {
        requisitions.register("testButtonClick", requestButtonClick);
        const component = shallow(
            <Button requestType="testButtonClick">
                Test button
            </Button>,
        );
        expect(component).toBeTruthy();
        expect(component.text()).toEqual("Test button");
        expect(clicked).toEqual(false);
        const click = (component.props() as IButtonProperties).onClick;
        void act(() => {
            return click?.(eventMock, { id: "1" });
        }).then(() => {
            sleepAsync((result: boolean) => {
                expect(result).toEqual(true);
                done();
            });
        });
    });

    it("Test button click", async () => {
        const innerRef = React.createRef<HTMLButtonElement>();
        const component = mount(
            <Button innerRef={innerRef} onClick={buttonClick}>
                Test button
            </Button>,
        );

        expect(innerRef.current).not.toEqual(document.activeElement);

        expect(component).toBeTruthy();
        expect(component.text()).toEqual("Test button");

        const click = (component.props() as IButtonProperties).onClick;
        await act(() => {
            click?.(eventMock, { id: "1" });
        });
        expect(clicked).toEqual(true);

        const onMouseDown = (component.props() as IButtonProperties).onMouseDown;
        await act(() => {
            onMouseDown?.(eventMock as React.MouseEvent, { id: "1" });
        });
        expect(innerRef.current).not.toEqual(document.activeElement);

        component.unmount();
    });

    it("Test checkbox output (Snapshot)", () => {
        const component = mount(
            <div>
                <Container>
                    <Button id="button1" caption="Normal Button" />
                    <Button id="button2" caption="Disabled Button" disabled />
                </Container>
                <Container>
                    <Button id="button3" orientation={Orientation.TopDown}>
                        Button<h2>with content</h2>
                    </Button>
                    <Button id="button4">
                        <Image src={image} />
                        Image Button
                    </Button>
                </Container>
                <Button
                    id="button5"
                    caption="Large Button"
                    size={ComponentSize.Big}
                />
                <Container>
                    <Button round>M</Button>
                    <Button round>
                        <Image src={image} />
                    </Button>
                </Container>
            </div>,
        );
        expect(snapshotFromWrapper(component)).toMatchSnapshot();

        component.unmount();
    });
});
