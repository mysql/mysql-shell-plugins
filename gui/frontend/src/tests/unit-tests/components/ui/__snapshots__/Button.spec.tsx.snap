// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component tests Test checkbox output (Snapshot) 1`] = `
<div>
  <Container
    orientation="row"
  >
    <div
      className="msg container fixedScrollbar"
      style={
        Object {
          "alignItems": undefined,
          "flexDirection": "row",
          "flexWrap": undefined,
          "justifyContent": undefined,
        }
      }
    >
      <Button
        caption="Normal Button"
        id="button1"
      >
        <button
          caption="Normal Button"
          className="msg button"
          id="button1"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          Normal Button
        </button>
      </Button>
      <Button
        caption="Disabled Button"
        disabled={true}
        id="button2"
      >
        <button
          caption="Disabled Button"
          className="msg button"
          disabled={true}
          id="button2"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          Disabled Button
        </button>
      </Button>
    </div>
  </Container>
  <Container
    orientation="row"
  >
    <div
      className="msg container fixedScrollbar"
      style={
        Object {
          "alignItems": undefined,
          "flexDirection": "row",
          "flexWrap": undefined,
          "justifyContent": undefined,
        }
      }
    >
      <Button
        id="button3"
        orientation="column"
      >
        <button
          className="msg button"
          id="button3"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": "column",
            }
          }
          tabIndex={0}
        >
          Button
          <h2>
            with content
          </h2>
        </button>
      </Button>
      <Button
        id="button4"
      >
        <button
          className="msg button"
          id="button4"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          <Image
            disabled={false}
            src="close.svg"
          >
            <img
              className="msg image"
              src="close.svg"
            />
          </Image>
          Image Button
        </button>
      </Button>
    </div>
  </Container>
  <Button
    caption="Large Button"
    id="button5"
    size="big"
  >
    <button
      caption="Large Button"
      className="msg button big"
      id="button5"
      onMouseDown={[Function]}
      style={
        Object {
          "flexDirection": undefined,
        }
      }
      tabIndex={0}
    >
      Large Button
    </button>
  </Button>
  <Container
    orientation="row"
  >
    <div
      className="msg container fixedScrollbar"
      style={
        Object {
          "alignItems": undefined,
          "flexDirection": "row",
          "flexWrap": undefined,
          "justifyContent": undefined,
        }
      }
    >
      <Button
        round={true}
      >
        <button
          className="msg button round"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          M
        </button>
      </Button>
      <Button
        round={true}
      >
        <button
          className="msg button round"
          onMouseDown={[Function]}
          style={
            Object {
              "flexDirection": undefined,
            }
          }
          tabIndex={0}
        >
          <Image
            disabled={false}
            src="close.svg"
          >
            <img
              className="msg image"
              src="close.svg"
            />
          </Image>
        </button>
      </Button>
    </div>
  </Container>
</div>
`;
